authority/provisioner/webhook.go

```diff
@@ -0,0 +1,204 @@
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"text/template"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/smallstep/certificates/templates"
+	"github.com/smallstep/certificates/webhook"
+	"go.step.sm/crypto/sshutil"
+	"go.step.sm/crypto/x509util"
+	"go.step.sm/linkedca"
+)
+
+var ErrWebhookDenied = errors.New("webhook server did not allow request")
+
+type WebhookSetter interface {
+	SetWebhook(string, any)
+}
+
+type WebhookController struct {
+	client       *http.Client
+	webhooks     []*Webhook
+	certType     linkedca.Webhook_CertType
+	TemplateData WebhookSetter
+}
+
+// Enrich fetches data from remote servers and adds returned data to the
+// templateData
+func (wc *WebhookController) Enrich(req *webhook.RequestBody) error {
+	if wc == nil {
+		return nil
+	}
+	for _, wh := range wc.webhooks {
+		if wh.Kind != linkedca.Webhook_ENRICHING.String() {
+			continue
+		}
+		if !wc.isCertTypeOK(wh) {
+			continue
+		}
+		resp, err := wh.Do(wc.client, req, wc.TemplateData)
+		if err != nil {
+			return err
+		}
+		if !resp.Allow {
+			return ErrWebhookDenied
+		}
+		wc.TemplateData.SetWebhook(wh.Name, resp.Data)
+	}
+	return nil
+}
+
+// Authorize checks that all remote servers allow the request
+func (wc *WebhookController) Authorize(req *webhook.RequestBody) error {
+	if wc == nil {
+		return nil
+	}
+	for _, wh := range wc.webhooks {
+		if wh.Kind != linkedca.Webhook_AUTHORIZING.String() {
+			continue
+		}
+		if !wc.isCertTypeOK(wh) {
+			continue
+		}
+		resp, err := wh.Do(wc.client, req, wc.TemplateData)
+		if err != nil {
+			return err
+		}
+		if !resp.Allow {
+			return ErrWebhookDenied
+		}
+	}
+	return nil
+}
+
+func (wc *WebhookController) isCertTypeOK(wh *Webhook) bool {
+	if wc.certType == linkedca.Webhook_ALL {
+		return true
+	}
+	if wh.CertType == linkedca.Webhook_ALL.String() || wh.CertType == "" {
+		return true
+	}
+	return wc.certType.String() == wh.CertType
+}
+
+type Webhook struct {
+	ID                   string `json:"id"`
+	Name                 string `json:"name"`
+	URL                  string `json:"url"`
+	Kind                 string `json:"kind"`
+	DisableTLSClientAuth bool   `json:"disableTLSClientAuth,omitempty"`
+	CertType             string `json:"certType"`
+	Secret               string `json:"-"`
+	BearerToken          string `json:"-"`
+	BasicAuth            struct {
+		Username string
+		Password string
+	} `json:"-"`
+}
+
+func (w *Webhook) Do(client *http.Client, reqBody *webhook.RequestBody, data any) (*webhook.ResponseBody, error) {
+	tmpl, err := template.New("url").Funcs(templates.StepFuncMap()).Parse(w.URL)
+	if err != nil {
+		return nil, err
+	}
+	buf := &bytes.Buffer{}
+	if err := tmpl.Execute(buf, data); err != nil {
+		return nil, err
+	}
+	url := buf.String()
+
+	switch tmpl := data.(type) {
+	case x509util.TemplateData:
+		reqBody.Token = tmpl[x509util.TokenKey]
+	case sshutil.TemplateData:
+		reqBody.Token = tmpl[sshutil.TokenKey]
+	}
```

Comment:
Although I insisted on this before, I later realized that these tokens might be problematic for a couple of reasons:
1. A K8sSA token can be reused multiple times. The webhook can misuse it to get fake certificates.
2. A webhook can misuse any other token to get its own certificate before responding.

I would comment those lines for the moment, with a note about the reasons, and we can discuss what we want to do with this.

Reply:
58be425